Allow multiple file uploads

pwkip opened the pull request:
I created this pull request following our discussion here: https://wordpress.org/support/topic/allow-upload-of-multiple-files-with-file-my-files-multiple/

This pull requests introduces the `multiple` option to the `[file]` tag. Use it like this:

```
[file my-files multiple]
[submit]
```

Since one file field can now have multiple files attached to it, I had to make some other structural changes, which other plugin developers might need to be aware of.

1. `$submission->uploaded_files` and `$submission->uploaded_files()` will now contain an array of paths, instead of a single path per name. For example: `['my-file' => ['path-to-file/file1.jpg', 'path-to-file/file2.jpg']]`. Even if there is only one file or if you omit the `multiple` option, it will still return an array with one path.
2. Same thing happens for `$submission->posted_data`. Instead of returning an md5 hash for each file field, it will return an array of md5 hashes for each file field.

Users of the plugin need to be aware of the following if they use `[file my-files multiple]`:

1. in the attachments field they can simply input `[my-files]`, this will attach all files to the email.
2. In the email body, `[my-files]` will be replaced by a comma-separated list of the uploaded file names.

@takayukister If you like this idea and would like to include it in the plugin, please let me know and I will make sure to run some additional tests.

takayukister commented:
To get this patch merged, at least, you need to make these points clear:

* Input validation. Such as, what type of validation error should be returned when a portion of files doesn't meet the requirements of the [form-tag options](https://contactform7.com/file-uploading-and-attachment/#adding-file-uploading-fields-in-your-form) (`filetypes` and `limit`)?
* How to cap the sum total size of uploaded files? What should be set in the default config? 
* Does the [configuration error](https://contactform7.com/configuration-errors/attachments-overweight/) need to be modified?
* Do all major browsers support it? If not, should some kind of fallback be provided?

pwkip commented:
Working on an update.

What is the minimum PHP version you require? It's not stated in your readme.txt file. I'm currently assuming PHP 5.6

takayukister commented:
If it is not stated then it is the same as [WordPress's requirement](https://wordpress.org/about/requirements/). Currently it is PHP 5.6.20.

pwkip commented:
I added some more validation, and a couple of extra options and error messages. An overview:

## Additional options
<img src="https://user-images.githubusercontent.com/1150196/89322388-6e227300-d684-11ea-8ae8-e763b801805b.png" width="300" />

- **multiple** - boolean, indicates whether multiple uploads are allowed or not. Default: false
- **min:** - integer, indicated the minimum number of files that must be uploaded if **multiple** is active. Default: 0
- **max:** - integer, indicated the maximum number of files that that are allowed if **multiple** is active. Default: 10
- **limit_total:** - integer (can be appended with B, MB, kB, ...), the max size of all files combined. Default: 1MB

## Additional error messages

<img src="https://user-images.githubusercontent.com/1150196/89323546-27ce1380-d686-11ea-89e3-95e66fa5988f.png" width="300" />


```PHP
'upload_files_too_large' => array(
	'description' => __( "Total size of the uploaded files is too large", 'contact-form-7' ),
	'default' => __( "The total size of the uploaded files is too large.", 'contact-form-7' )
),

'upload_too_many_files' => array(
	'description' => __( "Too many files files selected", 'contact-form-7' ),
	'default' => __( "You are not allowed to upload this many files.", 'contact-form-7' )
),

'upload_too_few_files' => array(
	'description' => __( "Too few files files selected", 'contact-form-7' ),
	'default' => __( "You must select more files.", 'contact-form-7' )
),
```
## Answers to your questions

#### Input validation. Such as, what type of validation error should be returned when a portion of files doesn't meet the requirements of the form-tag options (filetypes and limit)?

See validation errors above. Also see the form tester (below)

#### How to cap the sum total size of uploaded files? What should be set in the default config?

I have added the `limit_total` option to handle this.

#### Does the configuration error need to be modified?

I didn't test this, but I don't think anything should change. Seems fine to me. All files will be filtered individually, so you code calculating the total size should keep working.

#### Do all major browsers support it? If not, should some kind of fallback be provided?

`multiple` is supported by all modern browsers and IE10+ according to https://www.w3schools.com/tags/att_input_multiple.asp

Some problems with built-in Android browsers on android versions 4.0 and lower (pre-2011). https://caniuse.com/#feat=input-file-multiple

As this seems to only apply to android versions released almost 10 years ago now, I would say it's safe to say that browser support is sufficient to implement it without an extensive fallback. Of course there is always the default browser fallback: if the browser does not support the multiple attribute, the user will only be able to upload 1 file. This should work without a problem and without any other unwanted side-effects.

## Test tool

I have activated this fork on a form-tester page: https://conditional-fields-cf7.bdwm.be/form-tester/?test-form=multiple+file+uploads

You can modify the form here in real time and see a preview of the email that would be generated. (Note that the attachment field does not work yet in the test tool, but I have tested attachments with real forms and it works)

pwkip commented:
@takayukister Are there any additional actions you would like me to take?

takayukister commented:
I still see it as that risks surpass benefits so don't feel like including it as Contact Form 7's core feature. I think it's better to be implemented as a separate extension plugin.

pwkip commented:
Thanks for your reply Takayuki, I will create a separate plugin for this.
Could you list the main concerns you have? Then I will make sure to address them properly in my extension plugin.

pwkip commented:
For anyone stumbling on this thread. This PR had been implemented in Conditional Fields Pro
https://conditional-fields-cf7.bdwm.be/multifile/
